### PR TITLE
Add extractors

### PR DESCRIPTION
**Pull request by SteamRiC**

### CODE REVIEW

File: lib/salestation/web/extractors.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    module Extractors
+      class InputExtractor
+        include Deterministic
+
+        def initialize(&block)
+          @block = block
+        end
+
+        def call(sinatra_request)
+          @block.call(sinatra_request)
```
Comment (indrekj) [inline lib/salestation/web/extractors.rb]: rack_request? Is there anything sinatra specific?

File: salestation.gemspec [hunk 2]
```diff
@@ -4,7 +4,7 @@ $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
 
 Gem::Specification.new do |spec|
   spec.name          = "salestation"
-  spec.version       = "0.2.1"
+  spec.version       = "0.2.2"
```
Comment (indrekj) [inline salestation.gemspec]: feature bump makes more sense here

File: spec/salestation/web/extractors/body_param_extractor_spec.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
+require 'spec_helper'
+
+sinatra_request = Class.new do
```
Comment (indrekj) [inline spec/salestation/web/extractors/body_param_extractor_spec.rb]: don't be specific to sinatra

File: README.md [hunk 4]
```diff
@@ -75,6 +75,44 @@ Salestation allows and recommends you to define your own custom errors. This is
...
+    .merge(ConstantInput[name: name])
+
+  process extractor do |request|
+    HelloUser.call(request)
+      .map(Responses.to_ok)
+  end
+end
+```
+
+```ruby
```
Comment (indrekj) [inline README.md]: might as well put these two ruby blocks together if you don't write anything else here

File: README.md [hunk 5]
```diff
@@ -75,6 +75,44 @@ Salestation allows and recommends you to define your own custom errors. This is
...
+### Using Extractors
+
+Salestation provides extractors to fetch parameters from the request and pass them to the chain.
+Available extractors are `BodyParamExtractor`, `QueryParamExtractor`, `ConstantInput`.
+
+Define a route
+
+```ruby
+post '/hello/:name' do |name|
+  extractor = BodyParamExtractor[:age]
```
Comment (indrekj) [inline README.md]: this constant is available or do we need to include, prefix anything?

Comment (SteamRiC) [inline README.md]: Adding include to the example. Then it will be available

File: lib/salestation/web/extractors.rb (hunk 1 above)
Comment (SteamRiC) [inline lib/salestation/web/extractors.rb]: Nope

**Comment by SteamRiC** [comment]
Updated

File: README.md [hunk 6]
```diff
@@ -75,6 +75,44 @@ Salestation allows and recommends you to define your own custom errors. This is
   })
 ```
 
+### Using Extractors
+
+Salestation provides extractors to fetch parameters from the request and pass them to the chain.
+Available extractors are `BodyParamExtractor`, `QueryParamExtractor`, `ConstantInput`.
```
Comment (urmastalimaa) [inline README.md]: HeadersExtractor is also provided

Comment (urmastalimaa) [inline README.md]: Document also that the extractors can be merged and what the merge semantics are (next extractor overwrites 
previous value if same key).

File: README.md [hunk 7]
```diff
@@ -75,6 +75,44 @@ Salestation allows and recommends you to define your own custom errors. This is
...
+
+Define a route
```
Comment (urmastalimaa) [inline README.md]: `coercions` are undocumented

**Comment by SteamRiC** [comment]
Updated

File: README.md [hunk 8]
```diff
@@ -78,16 +78,22 @@ Salestation allows and recommends you to define your own custom errors. This is
 ### Using Extractors
 
 Salestation provides extractors to fetch parameters from the request and pass them to the chain.
-Available extractors are `BodyParamExtractor`, `QueryParamExtractor`, `ConstantInput`.
+Available extractors are `BodyParamExtractor`, `QueryParamExtractor`, `ConstantInput`, `HeadersExtractor`.
+Multiple extractors can be merged together. If two or more extractors use the same key, the value will be from the last extractor in the merge chain.
+
+`coersions` can optionally be provided to `BodyParamExtractor` and `QueryParamExtractor`. These can be used to transform the values of the extracted parameters.
```
Comment (sviik) [inline README.md]: `coersions` -> `coercions`

**Comment by SteamRiC** [comment]
Updated

**Comment by indrekj** [comment]
!merge

**Pull request merged**